pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -178,13 +179,17 @@ def setup_environment(self) -> None:
         # start the other scripts
         if not self.cluster_environment.creates_children():
             self._call_children_scripts()
+            self._has_called_call_children_scripts = True
 
         # set the task idx
         self.task_idx = self.cluster_environment.local_rank()
 
         self.setup_distributed()
 
     def _call_children_scripts(self):
+        if os.getenv("PL_DDP_CREATED_CHILDREN", "0") == "1":
+            return
+
```

Comment:
it's never getting set?

Reply:
Thanks, I forgot to remove it. But do you prefer an env variable or attribute to track that Lightning created the children ?

Reply:
IMO, I prefer an env var which remove the need for a broadcast.

Reply:
no please, let's not introduce global variables. the entire reason why debugging DDP was such a hell until Lightning 1.3 was because of all these global variables scattered around and modified everywhere. I removed all of this and DDP is working fine now.

I have strong feelings about this because at the core, DDP is not at all complicated to understand and use with PyTorch, yet Lightning is doing such an incredibly good job at making it so unbelievably complicated.

Reply:
@tchaton what was wrong around capturing this in the cluster environment initialization instead of in the DDP plug-in?

Was it that the cluster environment wasn't available on all subprocesses? That felt much more self-contained

Reply:
@awaelchli - some of this complexity also snowballs because Lightning is covering for what I think torch distributed should be responsible for instead. Deadlock detection is almost always due to collective calls hanging, and debugging this is really painful right now in pytorch. I believe folks from the distributed side are looking more into this though - fyi @cbalioglu

Reply:
Hey @ananthsub,

1. `LOCAL_RANK` not being available,  will result to LightningEnvironment.creates_children() -> False and therefore rank 0 will create the processes.

2. os.environ is populated with LOCAL_RANK, and rank 0 contains `_call_children_scripts=False` and rank > 0 contains `_call_children_scripts=True`

Yes, the only way is to broadcast from rank 0 to the other processes.